### ISSUE TITLE
Inconsistent behaviour handling blank Field Name/Value when editing data query suggested values

### ISSUE THREAD

**Issue by DavidLiedke** [Cacti/cacti#1530]
Data Collection --> Data Queries --> for example: SNMP Interfaces --> Associated Graph Templates --> for example: In/Out Bits (64-bit Counters) --> Suggested Values - Graphs

If you enter a "Suggested Values" but no "Field Name" you get an white page.

![2018-04-12_10-00-17](https://user-images.githubusercontent.com/25581728/38664351-92adcd90-3e39-11e8-970a-5e447f1c689e.gif)

Cacti 1.1.37

**Comment by netniV** [comment]
Are you seeing any errors in your log files?

**Comment by DavidLiedke** [comment]
No errors in cacti & apache log.

**Comment by netniV** [comment]
No, the reason is down to the way the IF statements are checking values.  For the data source suggestions, it only checks that you've clicked add.  For the graph, it checks that you have a query plus a name and value.  Because of those additional checks, it never actually does anything or returns anything.

The following is a patch that should fix it, can you try it out?
```diff
diff --git a/data_queries.php b/data_queries.php
index ad3472d..6abe735 100644
--- a/data_queries.php
+++ b/data_queries.php
@@ -206,13 +206,23 @@ function form_save() {
                }
 
                header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . (empty($snmp_query_graph_id) ? get_request_var('id') : $snmp_query_graph_id) . '&snmp_query_id=' . get_request_var('snmp_query_id'));
-       } elseif (isset_request_var('svg_x') && !isempty_request_var('svg_text') && !isempty_request_var('svg_field')) {
+       } elseif (isset_request_var('svg_x')) {
                /* ================= input validation ================= */
                get_filter_request_var('id');
                get_filter_request_var('snmp_query_id');
                get_filter_request_var('graph_template_id');
                /* ==================================================== */
 
+               if  (isempty_request_var('svg_text')) {
+                       raise_message(38);
+                       header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
+                       return;
+               } elseif (isempty_request_var('svg_field')) {
+                       raise_message(37);
+                       header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
+                       return;
+               }
+
                /* suggested values -- graph templates */
                $sequence = get_sequence(0, 'sequence', 'snmp_query_graph_sv', 'snmp_query_graph_id=' . get_filter_request_var('id') . " AND field_name = " . db_qstr(get_nfilter_request_var('svg_field')));
 
@@ -240,6 +250,16 @@ function form_save() {
                get_filter_request_var('graph_template_id');
                /* ==================================================== */
 
+               if  (isempty_request_var('svds_text')) {
+                       raise_message(38);
+                       header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
+                       return;
+               } elseif (isempty_request_var('svds_field')) {
+                       raise_message(37);
+                       header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
+                       return;
+               }
+
                foreach ($_POST as $var => $val) {
                        if (preg_match('/^svds_([0-9]+)_text/i', $var, $matches)) {
                                /* ================= input validation ================= */
```
```diff
diff --git a/include/global_arrays.php b/include/global_arrays.php
index 2fe1c08..e039c24 100644
--- a/include/global_arrays.php
+++ b/include/global_arrays.php
@@ -219,6 +219,12 @@ $messages = array(
        36  => array(
                'message' => __('The file highlighted does not exist.  Please enter a valid file name.'),
                'type' => 'error'),
+       37 => array(
+               'message' => __('Suggested Field Name was not entered.  Please enter a field name and try again.'),
+               'type' => 'error'),
+       38 => array(
+               'message' => __('Suggested Value was not entered.  Please enter a suggested value and try again.'),
+               'type' => 'error'),
        'clog_invalid' => array(
                'message' => __('Log file specified is not a Cacti log or archive file.'),
                'type' => 'info'),
```

**Comment by netniV** [comment]
If this works for you, I'll submit it as a pull request.  This may even make it into the hotfix that is coming.

**Comment by DavidLiedke** [comment]
After the update from include/global_arrays.php i cant access my cacti installation anymore. I get a complete white page. 

In the cacti log i see following:
PHP Fatal error:  Call to undefined function get_rrdtool_version() in /srv/www/cacti/include/global_arrays.php on line 446

After restoring the old version all works fine.

**Comment by netniV** [comment]
Can you copy the global_arrays.php to a different folder and then re-apply the patch purely to that.  Then run php -l global_arrays.php and see whether it passes the validation?

It sounds like the patch didn't get applied properly.

**Comment by DavidLiedke** [comment]
I found the problem. I replaced the php files from your commit in the 1.2 branch. Between the version 1.1.37 and the 1.2 branch some other minor changes. These have caused the problem.

Now it works fine :)

**Comment by netniV** [comment]
Ah, yeah, that's why I posted the patch separately 👍 I didn't want to have you wait for 1.2 🕦 

(I actually had to repatch for 1.2 as it was slightly different).

**Comment by hron84** [comment]
I want to reopen this issue. After upgrading to the latest Cacti, I cannot set a suggested Data Source field/value pair in data query. Graph ones work. Does not matter if I checkbox the data source under the Associated Data Templates section or not. Btw, some tooltip would be useful what that checkbox does.

**Comment by netniV** [comment]
I believe that I resolved the issue with Suggested Names in #1617 and the header issue with #1618

**Comment by hron84** [comment]
@netniV could you please say a version number where the fixes are shipped?

**Comment by netniV** [comment]
1.2 but technically that isn't shipped. That's a development branch.

**Comment by hron84** [comment]
@netniV When can we expect fix for this? Technically this makes custom SNMP monitoring unusable.

**Comment by lichtl** [comment]
yes this is not resolved (I tried with and without below diff in data_queries). This is on a fresh installation (I only renamed the In/Out Bits that existed into "In/Out Bits (64-bit Counters) (1 minute)" .
![image](https://user-images.githubusercontent.com/1228155/44795169-5bef4e80-abaa-11e8-8c56-0f32a16943de.png)

![image](https://user-images.githubusercontent.com/1228155/44795187-64e02000-abaa-11e8-84a1-a4a935f55e50.png)

```
diff --git a/data_queries.php b/data_queries.php
index 2952947..936601f 100644
--- a/data_queries.php
+++ b/data_queries.php
@@ -213,6 +213,12 @@ function form_save() {
                get_filter_request_var('graph_template_id');
                /* ==================================================== */

+               if (isset_request_var('header') && get_nfilter_request_var('header') == 'false') {
+                       $header = '&header=false';
+               } else {
+                       $header = '';
+               }
+
                if  (isempty_request_var('svg_text')) {
                        raise_message(39);
                        header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
@@ -250,6 +256,12 @@ function form_save() {
                get_filter_request_var('graph_template_id');
                /* ==================================================== */

+               if (isset_request_var('header') && get_nfilter_request_var('header') == 'false') {
+                       $header = '&header=false';
+               } else {
+                       $header = '';
+               }
+
                if  (isempty_request_var('svds_text')) {
                        raise_message(39);
                        header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));

```

**Comment by lichtl** [comment]
ok managed to got it working by applying https://github.com/arjanoosting/cacti/commit/2ea03676625b704a20a8771372adf9b9706c15dd

```
git diff data_queries.php
diff --git a/data_queries.php b/data_queries.php
index 2952947..0394d36 100644
--- a/data_queries.php
+++ b/data_queries.php
@@ -213,6 +213,12 @@ function form_save() {
                get_filter_request_var('graph_template_id');
                /* ==================================================== */

+               if (isset_request_var('header') && get_nfilter_request_var('header') == 'false') {
+                       $header = '&header=false';
+               } else {
+                       $header = '';
+               }
+
                if  (isempty_request_var('svg_text')) {
                        raise_message(39);
                        header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
@@ -250,22 +256,28 @@ function form_save() {
                get_filter_request_var('graph_template_id');
                /* ==================================================== */

-               if  (isempty_request_var('svds_text')) {
-                       raise_message(39);
-                       header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
-                       return;
-               } elseif (isempty_request_var('svds_field')) {
-                       raise_message(38);
-                       header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
-                       return;
+               if (isset_request_var('header') && get_nfilter_request_var('header') == 'false') {
+                       $header = '&header=false';
+               } else {
+                       $header = '';
                }

                foreach ($_POST as $var => $val) {
-                       if (preg_match('/^svds_([0-9]+)_text/i', $var, $matches)) {
+                       if (preg_match('/^svds_([0-9]+)_x/i', $var, $matches)) {
                                /* ================= input validation ================= */
                                input_validate_input_number($matches[1]);
                                /* ==================================================== */

+                               if  (isempty_request_var('svds_' . $matches[1] . '_text')) {
+                                       raise_message(39);
+                                       header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_quer
+                                       return;
+                               } elseif (isempty_request_var('svds_' . $matches[1] . '_field')) {
+                                       raise_message(38);
+                                       header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_quer
+                                       return;
+                               }
+
                                $sequence = get_sequence(0, 'sequence', 'snmp_query_graph_rrd_sv', 'snmp_query_graph_id=' . get_request_var('id')  . ' AND data_template_id=' . $matches[1] . " AND f

                                $hash = get_hash_data_query(0, 'data_query_sv_data_source');
@@ -847,16 +859,16 @@ function data_query_item_edit() {
                                                                <?php print __('Field Name');?>
                                                        </td>
                                                        <td>
-                                                               <input id='svds_field' type='text' name='svds_<?php print $data_template['id'];?>_field' size='15'>
+                                                               <input type='text' class='svds_field ui-state-default ui-corner-all' id='svds_<?php print $data_template['id'];?>_field' name='svds_<
                                                        </td>
                                                        <td class='nowrap'>
                                                                <?php print __('Suggested Value');?>
                                                        </td>
                                                        <td>
-                                                               <input id='svds_text' type='text' name='svds_<?php print $data_template['id'];?>_text' size='60'>
+                                                               <input type='text' class='svds_text ui-state-default ui-corner-all' id='svds_<?php print $data_template['id'];?>_text' name='svds_<?p
                                                        </td>
                                                        <td>
-                                                               <input id='svds_x' type='button' name='svds_x' value='<?php print __esc('Add');?>' title='<?php print __('Add Data Source Name Sugges
+                                                               <input type='button' class='svds_x ui-button ui-corner-all ui-widget' id='svds_<?php print $data_template['id'];?>_x' name='svds_<?ph
                                                        </td>
                                                </tr>
                                        </table>
@@ -913,10 +925,12 @@ function data_query_item_edit() {
                });
        });

-       $('input[id="svds_x"]').click(function() {
-               var svds_text_name=$('#svds_text').attr('name');
-               var svds_field_name=$('#svds_field').attr('name');
-               var svds_x_name=$('#svds_x').attr('name');
+
+       $('input.svds_x').click(function() {
+               var $form = $(this).closest('form');
+               var svds_text_name=$form.find('input.svds_text').attr('name');
+               var svds_field_name=$form.find('input.svds_field').attr('name');
+               var svds_x_name=$form.find('input.svds_x').attr('name');
                var jSON = $.parseJSON('{ ' +
                        '"graph_template_id_prev":"'+graph_template_id_prev + '", ' +
                        '"action":"save", ' +
@@ -927,9 +941,10 @@ function data_query_item_edit() {
                        '"__csrf_magic":"'+csrfMagicToken+'", ' +
                        '"save_component_snmp_query_item":"1", ' +
                        '"snmp_query_id":"'+$('#snmp_query_id').val() + '", ' +
-                       '"'+svds_field_name+'":"'+$('#svds_field').val() + '", ' +
-                       '"'+svds_text_name+'":"'+$('#svds_text').val() + '", ' +
-                       '"'+svds_x_name+'":"Add" }');
+                       '"'+svds_field_name+'":"'+$form.find('input.svds_field').val() + '", ' +
+                       '"'+svds_text_name+'":"'+$form.find('input.svds_text').val() + '", ' +
+                       '"'+svds_x_name+'":"Add", ' +
+                       '"svds_x":"Add" }');

                $.post('data_queries.php', jSON).done(function(data) {
                        $('#main').html(data);

```

**Comment by janmaris** [comment]
Hi, I still see this issue in 1.1.38.
Can somebody explain how to solve this?
thanks.

**Comment by pf2** [comment]
> Hi, I still see this issue in 1.1.38.
> Can somebody explain how to solve this?
> thanks.

If you patched a generic 1.1.38 with few patches above try this little patch but use it only if you have on lines 253-261 this code:

```
253                 if  (isempty_request_var('svds_text')) {
254                         raise_message(39);
255                         header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
256                         return;
257                 } elseif (isempty_request_var('svds_field')) {
258                         raise_message(38);
259                         header('Location: data_queries.php?header=false&action=item_edit' . $header . '&id=' . get_request_var('id') . '&snmp_query_id=' . get_request_var('snmp_query_id'));
260                         return;
261                 }
```

and here is the patch
```
--- a/data_queries.php	2018-10-12 16:00:10.244168684 +0200
+++ b/data_queries.php	2018-10-12 16:01:25.507703246 +0200
@@ -927,6 +927,8 @@
 			'"__csrf_magic":"'+csrfMagicToken+'", ' +
 			'"save_component_snmp_query_item":"1", ' +
 			'"snmp_query_id":"'+$('#snmp_query_id').val() + '", ' +
+			'"svds_field":"'+$('#svds_field').val() + '", ' +
+			'"svds_text":"'+$('#svds_text').val() + '", ' +
 			'"'+svds_field_name+'":"'+$('#svds_field').val() + '", ' +
 			'"'+svds_text_name+'":"'+$('#svds_text').val() + '", ' +
 			'"'+svds_x_name+'":"Add" }');
```

**Comment by hron84** [comment]
@DavidLiedke as I see, there is no 1.2 release in the short future. Could you issue a bugfix release for this issue, because this really messes up the SNMP checks (and potentially others too, but I overuse this form for custom SNMP checks)?

**Comment by netniV** [comment]
Whilst @DavidLiedke was having some of the issues, ultimately, the only fixes for 1.1.38 are the patches that I provided above.  There is a pull request open with a fix (not sure if it's related to this) that is also for 1.1.38.

**Comment by lx1** [comment]
The patch by pf2 above works for me on 1.1.38.
